### PR TITLE
Extract `EngagementManager` and cover it with unit tests

### PR DESCRIPTION
**Pull request by wzieba**
### Description

This PR extracts `EngagementManager` to a separate unit and adds unit tests for it. 

The proposed unit test is based on manual `Thread#sleep` calls and is time-sensitive. It's due to usage of `java.utils.Timer` which runs a new thread internally - a thread we cannot control. Still, the unit test is written in a way to not be flaky (usages of `org.assertj.core.api.isCloseTo` assertions) and brings value. E.g., the current test suite covers fix added in #62 .

### How to test

Review and green light on CI should be just fine.

### CODE REVIEW

File: parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java
Comment (wzieba) [inline parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java:14]: Unit tests for this class are planned to be added in a seperate PR.

File: parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java
Comment (ParaskP7) [inline parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java:25]: Suggestion (💡): I suggest making this `@Nullable` and then checking that during `stop()`:

```java
    public void stop() {
        if (waitingTimerTask != null) {
            waitingTimerTask.cancel();
        }
        started = false;
    }
```

PS: Actually, I would even recommend for such refactor/test PRs to add any missing nullability annotation wherever possible going forward. 🤷

File: parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java
Comment (ParaskP7) [inline parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java:30]: Minor (🔍): This line exceed the usual `120` character limit guideline that we have on other projects, consider reformatting it to something more readable like:

```java
    public EngagementManager(
            ParselyTracker parselyTracker,
            Timer parentTimer,
            long intervalMillis,
            Map<String, Object> baseEvent,
            UpdateEngagementIntervalCalculator intervalCalculator
    ) {
```

File: parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java
Comment (ParaskP7) [inline parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java:55]: Suggestion (💡): Consider resolving/suppressing all the warnings related to this method:

1. Suppress: `Unchecked cast: 'java.lang.Object' to 'java.util.Map<java.lang.String,java.lang.Object>'`
2. Resolve: `Method invocation 'equals' may produce 'NullPointerException'`
3. Resolve: `Unboxing of 'baseMetadata.get("duration")' may produce 'NullPointerException'`

FYI: For no.2 there is an easy fix, just flit the `equals(...)`, for example:

**BEFORE**: baseEvent.get("url").equals(url)
**AFTER**: url.equals(baseEvent.get("url")

File: parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java
Comment (ParaskP7) [inline parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java:96]: Minor (🔍): Consider removing the redundant `(Map<String, Object>)` casting for `baseEventData`.

Warning: `Casting 'baseEventData' to 'Map<String, Object>' is redundant`

File: parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java
Comment (ParaskP7) [inline parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java:20]: 1. Warning (⚠️): I am not sure if this was done due to an oversight or by design but usually `ParselyTracker.PLog(...)` is used instead. Consider either removing this completely or using `ParselyTracker.PLog(...)` instead.
2. Question (❓): Now that I am seeing the `System.out.println(...)` logging, what about replacing the whole `ParselyTracker.PLog(...)` with the native Android logger, of course, and this way to be having a more fine-grained level of logging (`Log.d/i/e(...)`), as a separate issue/PR, wdyt? 🤔

File: parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt
Comment (ParaskP7) [inline parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt:20]: Minor (🔍): Consider moving this suppression closer to the warning, that is, instead of having it here for the class class, to only apply that to the `isCorrectEvent(...)` extension function.

File: parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt
Comment (ParaskP7) [inline parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt:141]: Minor (🔍): Since this is a `val` field and not a method or an inner class, for better visibility, consider moving that at the top, alongside the rest of the fields.

File: parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt
Comment (ParaskP7) [inline parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt:144]: Minor (🔍): Since this is a `private` function and not an inner class, for better visibility, consider moving that amongst the other private functions.

FYI: I would have moved that above the `isCorrectEvent(...)` extension method since it is always being used first to that during testing.

File: parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt
Comment (ParaskP7) [inline parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt:50]: Minor (🔍): Consider making this assertion more readable, just like you did with the test further below:

```kotlin
        assertThat(tracker.events[0]).isCorrectEvent(
            // Ideally: totalTime should be equal to DEFAULT_INTERVAL_MILLIS
            withTotalTime = { isCloseTo(DEFAULT_INTERVAL_MILLIS, withinPercentage(10)) },
            // Ideally: timestamp should be equal to System.currentTimeMillis() at the time of recording the event
            withTimestamp = { isCloseTo(timestamp, within(100L)) }
        )
```

File: parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt
Comment (ParaskP7) [inline parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt:59]: Minor (🔍): On the tests below you are using `20L` instead of `100L`. Consider changing that value to `20L` as well.

File: parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt
Comment (ParaskP7) [inline parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt:66]: Question (❓): Please help me understand how `start()` works on `EngagementManager` for multiple events. 🙏

FYI: I am asking that because, to my understanding, a new instance of `EngagementManager` is created with just one event, the `baseEvent`, and that is the only event needed to be enqueued. I am struggling a bit to understand the logic here and when we can have multiple of the same events queuing. 🤔

PS: This is related to 51a467463c5f0602100a436f76d21bed2f375db9 .

File: parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt
Comment (ParaskP7) [inline parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt:107]: Minor (🔍): Consider moving the `.hasEntrySatisfying("data")` check above the `.containsEntry("inc", 0L)` check to make it easier to follow the same logical order found within the `EngagementManager.doEnqueue(...)` method, that is, without having to jump up and down to verify test correctness.

```kotlin
        return containsEntry("action", "heartbeat")
            // Incremental will be always 0 because the interval is lower than 1s
            .hasEntrySatisfying("data") { data ->
                data as Map<String, Any>
                assertThat(data).hasEntrySatisfying("ts") { timestamp ->
                    timestamp as Long
                    assertThat(timestamp).withTimestamp()
                }.containsAllEntriesOf(testData.minus("ts"))
            }.containsEntry("inc", 0L)
            .hasEntrySatisfying("tt") { totalTime ->
                totalTime as Long
                assertThat(totalTime).withTotalTime()
            }
```

File: parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt
Comment (ParaskP7) [inline parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt:158]: Minor (🔍): Consider adding an empty line at the end of this test.

FYI: You can enable the `Ensure every saved file ends with a line break` option within your IDE.

PS: You will find that within `Editor` -> `General` -> `On Save` section at the bottom.

**Review by ParaskP7 — CHANGES_REQUESTED** [review]
👋 @wzieba !

I have reviewed and tested this PR as per the instructions, everything works as expected, another great testing job well done! 🌟 🧪 🌟 

But wow, that is a very convoluted test, I am very sorry you had to go through all this... 🫂 

-----

I have left a warning (⚠️). a couple of questions (❓), a few suggestions (💡) and some minor (🔍) comments for you to consider. I am going to NOT going to approve this PR for now, just because of this one warning (⚠️) comment, but other than that everything seems okay to me.

-----

`EXTRAS`
1. Suggestion (💡): I have noticed that during you refactor commits some of them wouldn't be able to build, just because it would require this other subsequent commit that you add later on. I would personally merge those 2 commits together and make it easier for the reviewer to understand that in order to (for example) extract a class, one would also need to make some methods more public (package private). 🤔
2. Suggestion (💡): In this PR, I am also seeing some additional commits that are not part of this PR. Consider rebasing you branch with the `main` one before open a PR for review. This will make it easier for reviewers to know exactly what need to be reviewed and avoid having them to figure out the out-of-scope commits.

File: parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java
Comment (ParaskP7) [inline parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java:8]: Suggestion (💡): Ah, I forgot to mention this in my review, what about making this a `Utils` class instead of instantiating it as an object. Maybe even a Kotlin utils class, wdyt? 🤔

File: parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java
Comment (wzieba) [reply parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java:30]: Sure, but I'd leave this, and other static code changes, to some linter. I'd prefer to not focus on this on reviews, as I'm afraid it'll shadow other, more functional comments. Do you have a suggestion for a linter/formatter tool for Java?

File: parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java
Comment (wzieba) [reply parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java:55]: I'd leave this to a different PR - I wanted this PR to be focused on extracting `EngagementManager` and adding minimal set of changes to make it testable.

File: parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java
Comment (wzieba) [reply parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java:20]: Thanks, I'll remove this log.

> System.out.println(...) logging, what about replacing the whole ParselyTracker.PLog(...) with the native Android logger,

We could do this, but at the same time, I don't see really a value in such change for this SDK. Could you elaborate, what would we gain by using the Android logger?

File: parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt
Comment (wzieba) [reply parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt:66]: When we call `start` on `EngagementManager`, the `java.utils.TimerTask` is scheduled inside `java.utils.Timer`. The next execution of the `TimerTask` will schedule a subsequent task.

So this test checks if subsequent tasks are executed and if they have correct parameters like e.g. updated total timestamp. Does this answer your questions @ParaskP7 ?

File: parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java
Comment (ParaskP7) [reply parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java:30]: > Sure, but I'd leave this, and other static code changes, to some linter. I'd prefer to not focus on this on reviews, as I'm afraid it'll shadow other, more functional comments.

Sure, this is just a minor one, feel free to skip those when needed. 👍 

> Do you have a suggestion for a linter/formatter tool for Java?

On most of our other Java-utilizing repos, we are and have been using `CheckStyle`, but then again we are not actively updating this tool on a regular basis, or at all for that matter, thus I am not sure. Having said that, `CheckStyle` indeed seems to be the best we can do for now, in only for consistency purpose... 🤷

File: parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java
Comment (ParaskP7) [reply parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java:20]: > Could you elaborate, what would we gain by using the Android logger?

Well, as I said above, a more fine-grained level of logging (`Log.d/i/e(...)`). Apart from that, I am not sure sure if using `System.out.println(...)` logging has some performance implication for apps that are using this SDK, or even if apps can remove this logging if they want to, with the native Android logging, I think the can, right... 🤷 🤔

File: parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt
Comment (wzieba) [reply parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt:141]: If that's okay, I'd leave this as it is. My reasoning: on top, we have declarations for the test subject and its dependencies. At the bottom: fakes and helper methods for test execution. `now` being `val` is only a syntax sugar to not add `()` in some places.

File: parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt
Comment (ParaskP7) [reply parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt:66]: Hmmm, yes, I understand how the `TimerTask` will schedule a subsequent task. I just don't understand when multiple events, mostly in production, could be scheduled, can you help me with that? 🙏 

```java
        engagementManager = new EngagementManager(this, timer, DEFAULT_ENGAGEMENT_INTERVAL_MILLIS, event, intervalCalculator);
        engagementManager.start();
```

I see the above happening on production, which on (for example) `startEngagement(...)` creates a new instance of `EngagementManager` with a specific single `event`. How is this manager is able to handle multiple events, when does this happens, this is what I don't understand @wzieba . 🤔

File: parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java
Comment (wzieba) [reply parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java:20]: Gotcha, thanks. I've added an issue for that: https://github.com/Parsely/parsely-android/issues/83 - I thought that `PLog` logs only if SDK has `debug` flag to `true`.

Comment (ParaskP7) [reply parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java:20]: Awesome, thanks for opening this issue @wzieba ! 🙇 

> I thought that `PLog` logs only if SDK has `debug` flag to `true`.

Wishful thinking... 😅 🤷

File: parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt
Comment (wzieba) [reply parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt:66]: >  How is this manager is able to handle multiple events, when does this happens, this is what I don't understand @wzieba . 🤔

When you say "multiple events" do you mean "multiple engagement sessions"? If yes, then the SDK cannot handle it.

Circling back to the original question:

> I am asking that because, to my understanding, a new instance of EngagementManager is created with just one event, the baseEvent, and that is the only event needed to be enqueued.

Yes, the `EngegementManager` is created with just one event, the `baseEvent`. But no - it's not the only event needed to be enqueued. After the initial `baseEvent` is enqueued, enqueuing of the next event is scheduled.

When interval time passes, the SDK, based on `baseEvent`, enqueues a new event with modifications done in `doEnqueue`. It modifies, e.g. `ts`, `inc` or `tt` parameters.

Comment (ParaskP7) [reply parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt:58]: > When you say "multiple events" do you mean "multiple engagement sessions"? If yes, then the SDK cannot handle it.

Good to know! 💯 

> Yes, the EngegementManager is created with just one event, the baseEvent. But no - it's not the only event needed to be enqueued. After the initial baseEvent is enqueued, enqueuing of the next event is scheduled.

👍 

> When interval time passes, the SDK, based on baseEvent, enqueues a new event with modifications done in doEnqueue. It modifies, e.g. ts, inc or tt parameters.

That explains it, and apologies for only knowing the basics of the SDK, trying to figure out the rest during reviews... 😊 

So, when `startEngagement(...)` gets triggered, it is always triggered with a `baseEvent`, and then, until `stopEngagement(...)` is called, the same `baseEvent` will be enqueued multiple time, based on the interval, with only some of its parameters being modified. If so, me repeating correctly what you already said, I then think I got the complete picture now, thanks so much for explaining. 🙇

File: parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java
Comment (wzieba) [reply parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java:8]: In next iterations, I'll inject some `Clock` to this class to fake the current time. Otherwise, test for this class will be time-sensitive. So I'll have to leave this as a class, not a static one.

File: parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java
Comment (wzieba) [reply parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java]: I've fixed the comments! After the second thought, it's probably not worth to introduce Java linter - we plan to use Kotlin anyway.

File: parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java
Comment (ParaskP7) [reply parsely/src/main/java/com/parsely/parselyandroid/UpdateEngagementIntervalCalculator.java:8]: Nice to know, thanks for the heads-up on that. 💯

File: parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java
Comment (ParaskP7) [reply parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java]: > I've fixed the comments! 

Checking @wzieba and thanks! 👀 

> After the second thought, it's probably not worth to introduce Java linter - we plan to use Kotlin anyway.

Yeaaa, maybe not... 🤷

**Review by ParaskP7 — APPROVED** [review]
Took it for another spin, everything LGTM now @wzieba , thanks again! 🚀

File: parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt
Comment (wzieba) [reply parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt:58]: > the same baseEvent will be enqueued multiple time, based on the interval, with only some of its parameters being modified

Yes, exactly this 🙌 !

**Comment by wzieba** [comment]
Thank you @ParaskP7 ! 🎉

**Comment by wzieba** [comment]
@ParaskP7 one question on the initial review:

> Suggestion (💡): In this PR, I am also seeing some additional commits that are not part of this PR. Consider rebasing you branch with the main one before open a PR for review.

Which additional commits did you have in mind?

**Comment by ParaskP7** [comment]
👋 @wzieba and thanks for the reply! 🙇 

Those [two below](https://github.com/Parsely/parsely-android/pull/82/commits) got me a bit confused:

image

🤷 

I remembered reviewing those two on this #81 PR, see bottom of this [commit list](https://github.com/Parsely/parsely-android/pull/81/commits):

image

🤷 

Does that help? 🙏

**Comment by wzieba** [comment]
I see, thanks! It's interesting though that those commits have different hashes. Normally, we shouldn't see those commits after merging `main`.

**Comment by ParaskP7** [comment]
Yeaaa, not sure what happened there, that's why, by default, I always recommend rebasing branches with latest `main` branch before opening a PR... 🤷

**Pull request merged**
